### PR TITLE
Abstract Base Class Implementation

### PR DESCRIPTION
**Pull request by sgasparyan**
Implemented Abstract Base Classes and deleted util.raiseNotDefined() wherever possible.

### CODE REVIEW

**Comment by eriq-augustine** [comment]
Connor will take a looks soon, but some meta notes:
 - Try for more meaningful PR names. If everyone followed this pattern, we would just have four non-descriptive PRs and would just have to guess which is which.
 - You commit messages also need to more meaningful. See my comments here: https://github.com/linqs/pacman/pull/21#issuecomment-521103471
 - Having small commits is typically good, but this is too much. A commit should not inherently be a single file, it should be a single fix/feature. In this case, the changes should either be in one commit, or broken up by some logical partition like package.

File: pacai/student/multiAgents.py [hunk 1]
```diff
@@ -87,14 +87,11 @@ def evaluationFunction(self, currentGameState, action):
...
-    """
 
-    def __init__(self, index):
-        super().__init__(index)
+    Method to Implement:
 
-    def getAction(self, gameState):
-        """
-        Returns the minimax action from the current gameState using self.treeDepth
+    - def getAction(self, gameState):
```
Comment (cfpryor) [inline pacai/student/multiAgents.py]: Remove dashes

File: pacai/student/multiAgents.py [hunk 2]
```diff
@@ -87,14 +87,11 @@ def evaluationFunction(self, currentGameState, action):
...
+        - Returns the minimax action from the current gameState using self.treeDepth
         and self.evaluationFunction.
 
         Here are some method calls that might be useful when implementing minimax.
```
Comment (cfpryor) [inline pacai/student/multiAgents.py]: Move this portion of the comment above "Methods to implement" and take indent in one

File: pacai/student/multiAgents.py [hunk 3]
```diff
@@ -111,45 +108,40 @@ def getAction(self, gameState):
...
-        # *** Your Code Here ***
-        util.raiseNotDefined()
 
 class AlphaBetaAgent(MultiAgentSearchAgent):
     """
     Your minimax agent with alpha-beta pruning (question 3)
+
+    Method to Implement: 
+
+    - def getAction(self, gameState):
```
Comment (cfpryor) [inline pacai/student/multiAgents.py]: Remove dashes

File: pacai/student/multiAgents.py [hunk 4]
```diff
@@ -111,45 +108,40 @@ def getAction(self, gameState):
...
     Your expectimax agent (question 4)
-    """
 
-    def __init__(self, index):
-        super().__init__(index)
-
-    def getAction(self, gameState):
-        """
-        Returns the expectimax action using self.treeDepth and self.evaluationFunction
+    - def getAction(self, gameState):
```
Comment (cfpryor) [inline pacai/student/multiAgents.py]: Remove Dashes

File: pacai/student/multiAgents.py [hunk 5]
```diff
@@ -111,45 +108,40 @@ def getAction(self, gameState):
...
     """
     Your expectimax agent (question 4)
-    """
 
-    def __init__(self, index):
-        super().__init__(index)
-
-    def getAction(self, gameState):
-        """
-        Returns the expectimax action using self.treeDepth and self.evaluationFunction
```
Comment (cfpryor) [inline pacai/student/multiAgents.py]: Make sure to be consistent with the comments. Add in Method to implement as you did before.

File: pacai/student/multiAgents.py [hunk 6]
```diff
@@ -165,24 +157,22 @@ def betterEvaluationFunction(currentGameState):
...
 
-    def __init__(self, index):
-        super().__init__(index)
-
-    def getAction(self, gameState):
-        """
-        Returns an action.
+    Method to Implement:
+    - def getAction(self, gameState):
+        - Returns an action.
```
Comment (cfpryor) [inline pacai/student/multiAgents.py]: Same here, remove dashes. Also, make sure to be consistent with periods (i know it was in before but the others don't have them you should change this)

File: pacai/student/multiAgents.py [hunk 7]
```diff
@@ -165,24 +157,22 @@ def betterEvaluationFunction(currentGameState):
...
     Your agent for the mini-contest
-    """
 
-    def __init__(self, index):
-        super().__init__(index)
-
-    def getAction(self, gameState):
-        """
-        Returns an action.
+    Method to Implement:
```
Comment (cfpryor) [inline pacai/student/multiAgents.py]: Add space between Method to implement and the method

File: pacai/student/multiAgents.py [hunk 8]
```diff
@@ -165,24 +157,22 @@ def betterEvaluationFunction(currentGameState):
...
+    def __init__(self, index):
+        super().__init__(index)
 
-        # *** Your Code Here ***
-        util.raiseNotDefined()
 
 # Abbreviation
 # TODO(eriq): Get rid of this abbreviation
 better = betterEvaluationFunction
+
```
Comment (cfpryor) [inline pacai/student/multiAgents.py]: Remove line

File: pacai/student/qlearningAgents.py [hunk 9]
```diff
@@ -23,109 +23,60 @@ class QLearningAgent(ReinforcementAgent):
...
-        You can initialize Q-values here...
-        """
 
-        super().__init__(index, **args)
+    Methods to Implement:
 
-    def getQValue(self, state, action):
-        """
-        Returns Q(state,action)
+    - def getQValue(self, state, action):
```
Comment (cfpryor) [inline pacai/student/qlearningAgents.py]: Remove dashes

File: pacai/student/qlearningAgents.py [hunk 10]
```diff
@@ -23,109 +23,60 @@ class QLearningAgent(ReinforcementAgent):
...
+        - Returns Q(state,action)
         Should return 0.0 if we never seen
```
Comment (cfpryor) [inline pacai/student/qlearningAgents.py]: Make line consistent in length (i.e. there does not need to be a break between this line and then next)

File: pacai/student/qlearningAgents.py [hunk 11]
```diff
@@ -23,109 +23,60 @@ class QLearningAgent(ReinforcementAgent):
...
-        """
 
-        """ YOUR CODE HERE """
-        util.raiseNotDefined()
-        """ END CODE """
-
-    def getValue(self, state):
-        """
-        Returns max_action Q(state,action)
+    - def getValue(self, state):
```
Comment (cfpryor) [inline pacai/student/qlearningAgents.py]: remove dashes

File: pacai/student/qlearningAgents.py [hunk 12]
```diff
@@ -23,109 +23,60 @@ class QLearningAgent(ReinforcementAgent):
...
+        - Returns max_action Q(state,action)
         where the max is over legal actions. Note that if
```
Comment (cfpryor) [inline pacai/student/qlearningAgents.py]: Move sentences around to fit better

File: pacai/student/qlearningAgents.py [hunk 13]
```diff
@@ -23,109 +23,60 @@ class QLearningAgent(ReinforcementAgent):
...
-        """
-
-        """ YOUR CODE HERE """
-        util.raiseNotDefined()
-        """ END CODE """
-
-    def getPolicy(self, state):
-        """
-        Compute the best action to take in a state. Note that if there
+    - def getPolicy(self, state):
```
Comment (cfpryor) [inline pacai/student/qlearningAgents.py]: REmove dashes

File: pacai/student/qlearningAgents.py [hunk 14]
```diff
@@ -23,109 +23,60 @@ class QLearningAgent(ReinforcementAgent):
...
 
-        """ YOUR CODE HERE """
-        util.raiseNotDefined()
-        """ END CODE """
+    - def getAction(self, state):
 
-    def getAction(self, state):
-        """
-        Compute the action to take in the current state. With
+        -         Compute the action to take in the current state. With
```
Comment (cfpryor) [inline pacai/student/qlearningAgents.py]: Remove dashes, and space between dash and Compute

File: pacai/student/qlearningAgents.py [hunk 15]
```diff
@@ -23,109 +23,60 @@ class QLearningAgent(ReinforcementAgent):
...
         probability self.epsilon, we should take a random action and
         take the best policy action otherwise. Note that if there are
         no legal actions, which is the case at the terminal state, you
         should choose None as the action.
 
         HINT: You might want to use util.flipCoin(prob)
         HINT: To pick randomly from a list, use random.choice(list)
-        """
 
         # Pick Action
```
Comment (cfpryor) [inline pacai/student/qlearningAgents.py]: I don't think we need this line along with the following two. Maybe think about this for a bit and decide if you would have wanted it.

File: pacai/student/qlearningAgents.py [hunk 16]
```diff
@@ -23,109 +23,60 @@ class QLearningAgent(ReinforcementAgent):
...
-
-        """
-        Description:
-        [Enter a description of what you did here.]
-        """
-
-        """ YOUR CODE HERE """
-        util.raiseNotDefined()
-        """ END CODE """
+
```
Comment (cfpryor) [inline pacai/student/qlearningAgents.py]: remove extra spaces and return

File: pacai/student/qlearningAgents.py [hunk 17]
```diff
@@ -23,109 +23,60 @@ class QLearningAgent(ReinforcementAgent):
...
-        util.raiseNotDefined()
-        """ END CODE """
+        
 
         return action
 
-    def update(self, state, action, nextState, reward):
-        """
-        The parent class calls this to observe a
+    - def update(self, state, action, nextState, reward):
```
Comment (cfpryor) [inline pacai/student/qlearningAgents.py]: Remove dashes

File: pacai/student/qlearningAgents.py [hunk 18]
```diff
@@ -23,109 +23,60 @@ class QLearningAgent(ReinforcementAgent):
...
+        - The parent class calls this to observe a
         state = action => nextState and reward transition.
         You should do your Q-Value update here
```
Comment (cfpryor) [inline pacai/student/qlearningAgents.py]: try and fit sentences better

File: pacai/student/qlearningAgents.py [hunk 19]
```diff
@@ -23,109 +23,60 @@ class QLearningAgent(ReinforcementAgent):
...
-        Description:
-        [Enter a description of what you did here.]
+        You can initialize Q-values here...
         """
 
-        """ YOUR CODE HERE """
-        util.raiseNotDefined()
-        """ END CODE """
+        super().__init__(index, **args)
+
```
Comment (cfpryor) [inline pacai/student/qlearningAgents.py]: Remove extra space

File: pacai/student/qlearningAgents.py [hunk 20]
```diff
@@ -23,109 +23,60 @@ class QLearningAgent(ReinforcementAgent):
...
         You should do your Q-Value update here
 
         NOTE: You should never call this function,
         it will be called on your behalf
-        """
+    """
 
+    def __init__(self, index, **args):
         """
-        Description:
```
Comment (cfpryor) [inline pacai/student/qlearningAgents.py]: Put back the Description portion ( i believe this was an area for the students to put in comments)

File: pacai/student/qlearningAgents.py [hunk 21]
```diff
@@ -170,6 +121,15 @@ class ApproximateQAgent(PacmanQAgent):
     You should only have to overwrite getQValue
     and update. All other QLearningAgent functions
     should work as is.
+
+    Methods to Implement:
+
+    - def getQValue(self, state, action):
```
Comment (cfpryor) [inline pacai/student/qlearningAgents.py]: dashes

File: pacai/student/searchAgents.py [hunk 22]
```diff
@@ -33,6 +33,35 @@ class CornersProblem(SearchProblem):
...
+    - def successorStates(self, state):
+        - Returns successor states, the actions they require, and a cost of 1.
+
+        As noted in search.py:
+        For a given state, this should return a list of triples, (successor, action, stepCost),
+        where 'successor' is a successor to the current state, 'action' is the action
+        required to get there, and 'stepCost' is the incremental
+        cost of expanding to that successor
+
+
```
Comment (cfpryor) [inline pacai/student/searchAgents.py]: Hmm, we may want to rewrite the second half (but for now its fine)

File: pacai/student/searchAgents.py [hunk 23]
```diff
@@ -223,6 +209,13 @@ class AnyFoodSearchProblem(PositionSearchProblem):
 
     You can use this search problem to help you fill in
     the findPathToClosestDot method.
+
+    Method to Implement:
+
+    - def isGoal(self, state):
+        - The state is Pacman's position. Fill this in with a goal test
+        that will complete the problem definition.
+        x, y = state
```
Comment (cfpryor) [inline pacai/student/searchAgents.py]: remove x,y=state

File: pacai/student/searchAgents.py [hunk 24]
```diff
@@ -223,6 +209,13 @@ class AnyFoodSearchProblem(PositionSearchProblem):
 
     You can use this search problem to help you fill in
     the findPathToClosestDot method.
+
+    Method to Implement:
+
+    - def isGoal(self, state):
```
Comment (cfpryor) [inline pacai/student/searchAgents.py]: dashes

File: pacai/student/searchAgents.py [hunk 25]
```diff
@@ -241,41 +234,24 @@ def __init__(self, gameState):
...
-        """
 
-        # *** Your Code Here ***
-        util.raiseNotDefined()
+    Method to Implement:
 
-    def getAction(self, state):
-        """
-        From game.py:
+    - def getAction(self, state):
```
Comment (cfpryor) [inline pacai/student/searchAgents.py]: dashes

File: pacai/student/searchAgents.py [hunk 26]
```diff
@@ -241,41 +234,24 @@ def __init__(self, gameState):
...
         The BaseAgent will receive a GameState and must return an action from
         game.Directions.{North, South, East, West, Stop}
-        """
 
-        # *** Your Code Here ***
-        util.raiseNotDefined()
-        return None
+    - def registerInitialState(self, state):
+        - This method is called before any moves are made.
+
```
Comment (cfpryor) [inline pacai/student/searchAgents.py]: remove space

File: pacai/student/valueIterationAgents.py [hunk 27]
```diff
@@ -9,6 +9,22 @@ class ValueIterationAgent(ValueEstimationAgent):
     (see mdp.py) on initialization and runs value iteration
     for a given number of iterations using the supplied
     discount factor.
+
+    Methods to Implement:
+
```
Comment (cfpryor) [inline pacai/student/valueIterationAgents.py]: dashes

File: pacai/student/valueIterationAgents.py [hunk 28]
```diff
@@ -9,6 +9,22 @@ class ValueIterationAgent(ValueEstimationAgent):
...
+    - def getQValue(self, state, action):
+        - The q-value of the state action pair
+        (after the indicated number of value iteration passes).
+        Note that value iteration does not
```
Comment (cfpryor) [inline pacai/student/valueIterationAgents.py]: spacing seems strange

File: pacai/student/valueIterationAgents.py [hunk 29]
```diff
@@ -22,6 +38,7 @@ def __init__(self, index, mdp, discountRate = 0.9, iters = 100):
             mdp.getPossibleActions(state)
             mdp.getTransitionStatesAndProbs(state, action)
             mdp.getReward(state, action, nextState)
+
```
Comment (cfpryor) [inline pacai/student/valueIterationAgents.py]: Remove extra line

**Comment by sgasparyan** [comment]
Hey Connor I just finished fixing all the mistakes you pointed out.

**Comment by eriq-augustine** [comment]
Sergey! I told you before, commits should be one feature, step, or fix.
They should be self contained, and each one should result in a consistent codebase that functions.
Stop committing single files at a time!!!
This pollutes the history.

**Comment by sgasparyan** [comment]
Just delete the extra lines and figured out how to stop committing every file separately.

**Comment by sgasparyan** [comment]
Hey Connor I am ready for another code review.

**Comment by eriq-augustine** [comment]
LGTM

**Pull request merged**